Add optional inner samplers to make ParentBasedSampler more customizable

mbakalov opened the pull request:
Fixes #1710.

## Changes

According to the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#parentbased) the `ParentBasedSampler` should support optional parameters to allow delegating sampling decision for cases when remote/local parent was/wasn't sampled.

Added a new constructor to `ParentBasedSampler` that takes 4 more `Samplers` as arguments.

The spec still doesn't say anything about linked spans, so I left that part in.

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Changes in public API reviewed - there is a public API change for the new ctor.

johnduhart commented on src/OpenTelemetry/Trace/ParentBasedSampler.cs:
```diff
@@ -37,6 +47,41 @@ public ParentBasedSampler(Sampler rootSampler)
+            Sampler remoteParentNotSampled,
+            Sampler localParentSampled,
+            Sampler localParentNotSampled)
+            : this(rootSampler)
```
The other language implementations I've seen for this sampler always have the fields populated, eg. `remoteParentSampled ` would be set to an instance of `AlwaysOnSampler`. This would simplify `ShouldSample` as null checks are no longer required.

* [Ruby](https://github.com/open-telemetry/opentelemetry-ruby/blob/b3fd39fa202a0168f7c48304a8ba105d3f39bc1d/sdk/lib/opentelemetry/sdk/trace/samplers.rb#L67-L75)
* [Go](https://github.com/open-telemetry/opentelemetry-go/blob/c066f15ed74fec52c4755014bfcf2bdfb8e0e661/sdk/trace/sampling.go#L170-L175)

johnduhart replied:
Based on those two implementations and my understanding of the specification ("Optional parameters"), I would say users should be allowed to specify a single override sampler, and not require that all 4 parameters or passed.

mbakalov replied:
Thx! I also thought about this, but wanted to avoid creating additional instances of AlwaysOn/AlwaysOff samplers and the delegating, assuming they will be overridden quite rarely. But not really sure how big of an issue it would be.

I do agree it makes it inconvenient to require passing all 4 in case someone only needed on (or likely two). Perhaps a better api like below?
````csharp
new ParentBasedSampler(new AlwaysOnSampler()) { RemoteParentNotSampled = new AlwaysOnSampler() }
```

johnduhart replied:
> Thx! I also thought about this, but wanted to avoid creating additional instances of AlwaysOn/AlwaysOff samplers and the delegating, assuming they will be overridden quite rarely. But not really sure how big of an issue it would be.

Fair point.

> I do agree it makes it inconvenient to require passing all 4 in case someone only needed one (or likely two). Perhaps a better api like below?

Using properties would allow for the samplers to be changed after instantiation. Maybe optional parameters instead?

CodeBlanch replied:
Tricky API. What about a builder? Something like...
```csharp
new ParentBasedSamplerBuilder()
.AddSampler(target: ParentBasedSamplerTarget.RemoteSpan, whenParentSampled: true, sampler: new AlwaysOnSampler())
.AddSampler(target: ParentBasedSamplerTarget.LocalSpan, whenParentSampled: false, sampler: new AlwaysOffSampler())
.Build();
```
🤷

mbakalov replied:
Yeah we need init-only properties 😄 

I was also thinking of doing "options":
```csharp
new ParentBasedSampler(
    new AlwaysOnSampler(), // root
    new ParentBasedSamplerOptions
    {
        RemoteParentNotSampled = new AlwaysOnSampler()
    }
);
```

Build I like the builder too.

We should probably also add ability to customize link-sampling behavior through these while we're at it.

cijothomas replied:
@mbakalov Are you still exploring using the Options way or Builder way?

mbakalov replied:
@cijothomas - I now think maybe the current version I have is good enough?
```csharp
public ParentBasedSampler(Sampler rootSampler)
{
}

public ParentBasedSampler(
	Sampler rootSampler,
	Sampler remoteParentSampled = null,
	Sampler remoteParentNotSampled = null,
	Sampler localParentSampled = null,
	Sampler localParentNotSampled = null)
	: this(rootSampler)
{
}
```

Allows to do this:
```csharp
var pb = new ParentBasedSampler(new AlwaysOnSampler(), localParentNotSampled: new AlwaysOnSampler());
```

Which seems ok?

Java has the same nullable parameters constructor and also has the builder, but I think that's probably because you can't do named arguments for optional parameters in Java.

cijothomas replied:
This looks good enough.

cijothomas replied:
@johnduhart @CodeBlanch Please comment if you see concerns with this. This looks reasonable to me.

CodeBlanch replied:
Should we add the default samplers into the XML comments? Eg:
```csharp
        /// <param name="remoteParentNotSampled">
        /// A <see cref="Sampler"/> to delegate sampling decision to in case of
        /// remote parent (<see cref="ActivityContext.IsRemote"/> == true) with <see cref="ActivityTraceFlags.Recorded"/> flag == false.
        /// Default: <see cref="AlwaysOffSampler" />.
        /// </param>
```

mbakalov replied:
Wouldn't hurt - added!

cijothomas commented:
@mbakalov Given we are very close to 1.0, I'll hold off from this PR for 1.0, so that we can take more time and review.
The changes are additive only, so should not be an issue. Thanks for being patient!

cijothomas commented:
> @mbakalov Given we are very close to 1.0, I'll hold off from this PR for 1.0, so that we can take more time and review.
> The changes are additive only, so should not be an issue. Thanks for being patient!

Will come back to this this week.

mbakalov commented:
Changed the API a bit to make additional samplers optional parameters in the new constructor.

The compiler's overload resolution prefers the first constructor when no other parameters are provided, so the existing code will still compile and work as expected.
```csharp
public ParentBasedSampler(Sampler rootSampler)
{
}

public ParentBasedSampler(
	Sampler rootSampler,
	Sampler remoteParentSampled = null,
	Sampler remoteParentNotSampled = null,
	Sampler localParentSampled = null,
	Sampler localParentNotSampled = null)
	: this(rootSampler)
{
}
```

We can do things like `var pb = new ParentBasedSampler(new AlwaysOnSampler(), localParentNotSampled: new AlwaysOnSampler());`, which I think is a nice API now.

I'd also done some benchmarking and surprisingly the version that always delegates to inner samplers didn't turn out to be slower at all! Looks like all the inner sampler calls are getting inlined anyway. I've no idea why the delegating version is even a tiny bit faster with a slightly smaller code size too - perhaps I am doing something wrong?
```
|                    Method |     Mean |    Error |   StdDev |   Median | Ratio | RatioSD | Code Size |
|-------------------------- |---------:|---------:|---------:|---------:|------:|--------:|----------:|
|     ShouldSample_RootOnly | 19.60 ns | 1.285 ns | 3.790 ns | 18.09 ns |  1.00 |    0.00 |     723 B |
| ShouldSample_NullChecking | 18.77 ns | 0.428 ns | 0.627 ns | 18.69 ns |  0.83 |    0.09 |     891 B |
|   ShouldSample_Delegating | 18.34 ns | 0.411 ns | 0.821 ns | 18.26 ns |  0.80 |    0.09 |     719 B |
```

Finally, I've kept the current behavior of "always sample if a linked activity was sampled". To retain current semantics, it comes before the "parent not sampled" delegation, which is probably unexpected for the user. Should we provide a `Sampler parenkLinkSampled` parameter as well perhaps?

Switching this back to "ready for review" for more feedback 😄

mbakalov commented:
Added and cleaned up the benchmarking code I used to [here](https://github.com/mbakalov/opentelemetry-dotnet/blob/9e52cd3c70b079648cbd252ca34cd5269e0b3ba4/test/Benchmarks/Trace/ParentBasedSamplerBenchmarks.cs#L79):
```csharp
[Benchmark(Baseline = true)]
public void ShouldSample_Original()
{
	this.sampler.ShouldSample_Original(this.remoteParentSampledParams);

	// this.sampler.ShouldSample_Original(this.localParentSampledParams);
	// this.sampler.ShouldSample_Original(this.remoteParentNotSampledParams);
	// this.sampler.ShouldSample_Original(this.localParentNotSampledParams);
}

[Benchmark]
public void ShouldSample_Delegating()
{
	this.sampler.ShouldSample(this.remoteParentSampledParams);

	// this.delegatingSampler.ShouldSample_Delegating(this.localParentSampledParams);
	// this.delegatingSampler.ShouldSample_Delegating(this.remoteParentNotSampledParams);
	// this.delegatingSampler.ShouldSample_Delegating(this.localParentNotSampledParams);
}
```

``` ini
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18363.1379 (1909/November2018Update/19H2)
Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.103
  [Host]     : .NET Core 3.1.12 (CoreCLR 4.700.21.6504, CoreFX 4.700.21.6905), X64 RyuJIT
  DefaultJob : .NET Core 3.1.12 (CoreCLR 4.700.21.6504, CoreFX 4.700.21.6905), X64 RyuJIT
```

```
|                  Method |     Mean |    Error |   StdDev |   Median | Ratio | RatioSD |
|------------------------ |---------:|---------:|---------:|---------:|------:|--------:|
|   ShouldSample_Original | 15.11 ns | 0.387 ns | 1.032 ns | 14.77 ns |  1.00 |    0.00 |
| ShouldSample_Delegating | 13.84 ns | 0.284 ns | 0.526 ns | 13.67 ns |  0.91 |    0.06 |
```

The `ShouldSample_Delegating` is the current suggested implementation that always delegates, while the `ShouldSample_Original` was the code we have in 1.0.1 currently, also repeated below. No idea why the delegating one is slightly faster!

```csharp
public SamplingResult ShouldSample_Original(in SamplingParameters samplingParameters)
{
	var parentContext = samplingParameters.ParentContext;
	if (parentContext.TraceId == default)
	{
		// If no parent, use the rootSampler to determine sampling.
		return this.rootSampler.ShouldSample(samplingParameters);
	}

	// If the parent is sampled keep the sampling decision.
	if ((parentContext.TraceFlags & ActivityTraceFlags.Recorded) != 0)
	{
		return new SamplingResult(SamplingDecision.RecordAndSample);
	}

	if (samplingParameters.Links != null)
	{
		// If any linked context is sampled keep the sampling decision.
		// TODO: This is not mentioned in the spec.
		// Follow up with spec to see if context from Links
		// must be used in ParentBasedSampler.
		foreach (var parentLink in samplingParameters.Links)
		{
			if ((parentLink.Context.TraceFlags & ActivityTraceFlags.Recorded) != 0)
			{
				return new SamplingResult(SamplingDecision.RecordAndSample);
			}
		}
	}

	// If parent was not sampled, do not sample.
	return new SamplingResult(SamplingDecision.Drop);
}
```

mbakalov commented:
@cijothomas - just noticed a comment about links. I can remove as part of this PR. Will make an update later today!

cijothomas commented:
> ve as part of this PR. Will

Could you do a standalone PR for that alone to make it easy for tracking purposes :)

cijothomas reviewed (approved):
Please add a line to Changelog.md as well

cijothomas commented:
Leaving unmerged for 1 more day to give other reviewers a chance to raise concerns.
